lib/utils/metrics.js

```diff
@@ -0,0 +1,146 @@
+'use strict';
+
+const assert = require('assert');
+const uuid = require('uuid');
+
+/** @typedef { import('bmutex').Lock } Lock */
+
+const metrics = exports;
+
+/**
+ * @typedef {Object} Honey
+ */
+
+class Event {
+  /**
+   *
+   * @param {string} name the name for the event
+   * @param {Event | Honey} parent a parent Event or a libhoney instance
+   */
+  constructor(name, parent) {
+    assert(name, 'Event name is required');
+    assert(parent, 'Parent event or libhoney instance is required.');
+    if (parent instanceof Event) {
+      this.honey = parent.honey;
+      this.parent = parent;
+      this.trace_id = parent.trace_id;
+    } else {
+      this.honey = parent;
+      this.parent = null;
+      this.trace_id = uuid.v4();
+    }
+
+    this.event = this.honey.newEvent();
+
+    this.span_id = uuid.v4();
+    this.startTime = process.hrtime();
+    this.timers = new Map();
+
+    this.addField('name', name);
+    this.addField('service_name', 'hsd');
+    this.addField('trace.span_id', this.span_id);
+    this.addField('trace.trace_id', this.trace_id);
+    if(this.parent) {
+      this.addField('trace.parent_id', this.parent.span_id);
+      this.addField('parent_id', this.parent.span_id);
+    }
+    this.event.timestamp = Date.now();
+  }
+
+  startTimer(name) {
+    if(name) {
+      this.timers.set(name, process.hrtime());
+    } else {
+      this.startTime = process.hrtime();
+    }
+  }
+
+  /**
+   * Measure how long it takes to get a lock
+   * @param {Lock} lock a bmutex Lock
+   * @returns {Promise<any>} a function to call to release the lock
+   */
+  async measureLockWait(lock) {
+    this.addField('preLockQueueLength', lock.jobs.length);
+    const preLockTime = process.hrtime();
+    const unlock = lock.lock();
+
+    this.addField('postLockQueueLength', lock.jobs.length);
+    this.addField(
+      'lockWaitTime',
+      metrics.timerToMs(process.hrtime(preLockTime))
```

Comment:
I'm not convinced this code will properly time the locks. Looking at some previous usage:

```javascript
const unlock = await this.fundLock.lock();
try {
  return await this._sendOpen(name, force, options);
} finally {
  unlock();
}
```

The locking code is async. Line 66 of this file (start of timer) does not await, so the elapsed time would always report 0. I think the promises get plumbed through properly though so awaiting or not, it shouldn't cause a logic issue, just a metrics correctness issue.

Reply:
yeah, it should be `await lock.lock();`